layer-atlas/src/main/java/com/layer/atlas/util/imagepopup/AtlasImagePopupActivity.java

```diff
@@ -103,6 +129,75 @@ public static void init(LayerClient layerClient) {
         MessagePartRegionDecoder.init(layerClient);
     }
 
+    @Override
+    public boolean onCreateOptionsMenu(Menu menu) {
+        getMenuInflater().inflate(R.menu.atlas_image_menu, menu);
+        return true;
+    }
+
+    @Override
+    public boolean onOptionsItemSelected(MenuItem item) {
+       if(item.getItemId() == R.id.action_save) {
+           if(checkPermission()) {
+               saveImageToGallery();
+           } else {
+               requestPermission();
+           }
+           return true;
+       }
+       return super.onOptionsItemSelected(item);
+    }
+
+    @Override
+    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
+        switch (requestCode) {
+            case WRITE_PERMISSION_REQUEST: {
+                if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED)
+                    saveImageToGallery();
+            }
+            return;
+        }
+    }
+
+    private void saveImageToGallery() {
+        String extractedID = UUID.randomUUID().toString();
+        List<String> path = mMessagePartId.getPathSegments();
+        if(path !=null && path.size()>2)
+            extractedID = path.get(path.size()-1);
+        MessagePartDecoder decoderFactory = new MessagePartDecoder();
+        boolean saveSuccess = true;
+        Bitmap bitmap = null;
+        try {
+            bitmap = decoderFactory.decode(this, mMessagePartId);
+        } catch (Throwable ex){
+            saveSuccess = false;
+        }
+
+        if(saveSuccess) {
+            saveSuccess = MediaStore.Images.Media.insertImage(
```

Comment:
create a simple method like private boolean insertImage(@Nullable Bitmap bitmapToSave){
return bitmap !=null && MediaStore.Images.Media.insertImage(...) != null} then depends on state 
Toast.makeText(this, insertImage()? R.string.atlas_save_media_success : R.string.atlas_save_media_error , Toast.LENGTH_SHORT).show(); I think this solution will be simpler and needs less code :)